pennylane/math/quantum.py

```diff
@@ -164,6 +169,167 @@ def marginal_prob(prob, axis):
+        tensor_like: (reduced) Density matrix of size ``(2**len(wires), 2**len(wires))``
+
+    **Example**
+
+    >>> x = np.array([[1, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]])
+    >>> partial_trace(x, wires=[0])
+    [[1.+0.j 0.+0.j]
+     [0.+0.j 0.+0.j]]
+
+
+    >>> x = tf.Variable([[1, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]], dtype=tf.complex128)
+    >>> partial_trace(x, wires=[1])
+    tf.Tensor(
+    [[1.+0.j 0.+0.j]
+     [0.+0.j 0.+0.j]], shape=(2, 2), dtype=complex128)
+
+    >>> y = [[0.5, 0, 0.5, 0], [0, 0, 0, 0], [0.5, 0, 0.5, 0], [0, 0, 0, 0]]
+    >>> partial_trace(y, wires=[1])
+    [[0.5+0.j 0.5+0.j]
+     [0.5+0.j 0.5+0.j]]
+
+    >>> partial_trace(y, wires=[0])
+    [[1.+0.j 0.+0.j]
+     [0.+0.j 0.+0.j]]
+    """
+    # Dimension and reshape
+    shape = density_matrix.shape[0]
+    num_wires = int(np.log2(shape))
+    rho_dim = 2 * num_wires
+
+    density_matrix = np.reshape(density_matrix, [2] * 2 * num_wires)
+
+    # Kraus operator for partial tracee
+    kraus = cast(np.eye(2), dtype="complex128")
+    kraus = np.reshape(kraus, (2, 1, 2))
+    kraus_dagger = np.asarray([np.conj(np.transpose(k)) for k in kraus])
+
+    kraus = convert_like(kraus, density_matrix)
+    kraus_dagger = convert_like(kraus_dagger, density_matrix)
+
+    # For loop over wires
+    for target_wire in wires:
+        # Tensor indices of density matrix
+        state_indices = ABC[:rho_dim]
+
+        # row indices of the quantum state affected by this operation
+        row_wires_list = [target_wire]
+        row_indices = "".join(ABC_ARRAY[row_wires_list].tolist())
+
+        # column indices are shifted by the number of wires
+        col_wires_list = [w + num_wires for w in row_wires_list]
+        col_indices = "".join(ABC_ARRAY[col_wires_list].tolist())
+
+        # indices in einsum must be replaced with new ones
+        num_partial_trace_wires = 1
+        new_row_indices = ABC[rho_dim : rho_dim + num_partial_trace_wires]
+        new_col_indices = ABC[
+            rho_dim + num_partial_trace_wires : rho_dim + 2 * num_partial_trace_wires
+        ]
+
+        # index for summation over Kraus operators
+        kraus_index = ABC[
+            rho_dim + 2 * num_partial_trace_wires : rho_dim + 2 * num_partial_trace_wires + 1
+        ]
+
+        # new state indices replace row and column indices with new ones
+        new_state_indices = functools.reduce(
+            lambda old_string, idx_pair: old_string.replace(idx_pair[0], idx_pair[1]),
+            zip(col_indices + row_indices, new_col_indices + new_row_indices),
+            state_indices,
+        )
+
+        # index mapping for einsum, e.g., 'iga,abcdef,idh->gbchef'
+        einsum_indices = (
+            f"{kraus_index}{new_row_indices}{row_indices}, {state_indices},"
+            f"{kraus_index}{col_indices}{new_col_indices}->{new_state_indices}"
+        )
+
+        density_matrix = np.einsum(einsum_indices, kraus, density_matrix, kraus_dagger)

```

Comment:
It might be worth investigating how this could be done with `np.tensordot` instead. Even if there's a small performance decrease, the increased readability and maintainability of `tensordot` might make it worth it to use instead.

Reply:
Yes I agree that the implementation could be different more readable and also more efficient, it could a future point to tackle but for the moment we will use `einsum`.